KUBERNETES.md

```diff
@@ -0,0 +1,180 @@
+List of possible operations and corresponding JSON specification can be found below.
+
+The path to the file is found in the `$KUBERNETES_OPERATIONS_PATH` environment variable.
+
+> TODO: bash examples
+
+## Operations
+
+### Create
+
+* `operation` — specifies an operation's type.
+    * `CreateOrUpdate` — accept a Kubernetes object.
+      It retrieves an object, and if it already exists, computes a JSON Merge Patch and applies it (will not update .status field).
+      If it does not exist, we create the object.
+    * `Create` — will fail if an object already exists
+    * `CreateIfNotExists` — create an object if such an object does not already
+      exist by namespace/name.
+* `object` — full object specification including "apiVersion", "kind" and all necessary metadata.
+
+#### Example
+
+```json
+{
+  "operation": "CreateOrUpdate",
+  "object": {
+    "apiVersion": "apps/v1",
+    "kind": "DaemonSet",
+    "metadata": {
+      "name": "flannel",
+      "namespace": "d8-flannel"
+    },
+    "spec": {
+      "selector": {
+        "matchLabels": {
+          "app": "flannel"
+        }
+      },
+      "template": {
+        "metadata": {
+          "labels": {
+            "app": "flannel",
+            "tier": "node"
+          }
+        },
+        "spec": {
+          "containers": [
+            {
+              "args": [
+                "--ip-masq",
+                "--kube-subnet-mgr"
+              ],
+              "image": "flannel:v0.11",
+              "name": "kube-flannel",
+              "securityContext": {
+                "privileged": true
+              }
+            }
+          ],
+          "hostNetwork": true,
+          "imagePullSecrets": [
+            {
+              "name": "registry"
+            }
+          ],
+          "terminationGracePeriodSeconds": 5
+        }
+      },
+      "updateStrategy": {
+        "type": "RollingUpdate"
+      }
+    }
+  }
+}
+```
+
+### Delete
+
+* `operation` — specifies an operation's type. Deletion types map directly to Kubernetes
+  DELETE's [`propagationPolicy`](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/).
+  * `Delete` — foreground deletion. Hook will block the queue until the referenced object and all its descendants are deleted.
+  * `DeleteInBackground` — will delete the referenced object immediately. All its descendants will be removed by Kubernetes'
+    garbage collector.
+  * `DeleteNonCascading` — will delete the referenced object immediately, and orphan all its descendants.
+  apiVersion from `discovery` client.
+* `apiVersion` — optional field that specifies object's apiVersion. If not present, we'll use preferred apiVersion
+  for the given kind.
+* `kind` — object's Kind.
+* `namespace` — object's namespace. If empty, implies operation on a cluster-level resource.
+* `name` — object's name.
+* `subresource` — e.g., `status`.
```

Comment:
Do we really need to delete subresources? What will happen if we delete the status of a Deployment, for example?

Reply:
Makes sense.

Reply:
The API allows us to send a delete request to a subresource. We aren't enforcing any other constraints.